### PR TITLE
Add Microsoft.ML.GenAI.Phi, test package and sample project.

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

#7169

### CODE REVIEW

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.ML.Tokenizers;
+
+public class Phi3Tokenizer : Tokenizer
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs:9]: The `Phi3Tokenizer` can be replaced by `SentencePieceTokenizer` once #7158 is implemented.

File: src/Microsoft.ML.Tokenizers/Tokenizer.cs [hunk 2]
```diff
@@ -527,6 +527,24 @@ public static Tokenizer CreateTiktokenForEncoding(string encodingName, IReadOnly
             bool addEndOfSentence = false)
             => CreateCodeGen(vocabStream, mergesStream, addPrefixSpace, addBeginOfSentence, addEndOfSentence);
 
+        public static CodeGen CreatePhi2(
```
Comment (LittleLittleCloud) [inline src/Microsoft.ML.Tokenizers/Tokenizer.cs:530]: @tarekgh Can you review this change in tokenizer package

File: src/Microsoft.ML.Tokenizers/Tokenizer.cs [hunk 3]
```diff
@@ -527,6 +527,22 @@ public static Tokenizer CreateTiktokenForEncoding(string encodingName, IReadOnly
...
+            string folder,
```
Comment (tarekgh) [inline src/Microsoft.ML.Tokenizers/Tokenizer.cs:531]: I am not getting why you need this? can't you do that in the calling site in a helper method? Any API we expose need to be carefully evaluated before exposing it. Also, I am planning to expose something for Ph3 when fixing the special tokens issue. 

FYI, I am currently refactoring a lot of stuff in the code which we'll visit all calling sites and update them.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.Tokenizers/Tokenizer.cs:531]: I thought this is the place  where I can add tokenizer helper method to? If not I can move it towards `phi` project.

Comment (tarekgh) [reply src/Microsoft.ML.Tokenizers/Tokenizer.cs:531]: Please move it to your consuming project. Thanks!

Comment (LittleLittleCloud) [reply src/Microsoft.ML.Tokenizers/Tokenizer.cs]: Removed

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs (hunk 1 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs]: #7158 has done

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+{
+    private const string SystemSymbol = "<|system|>";
+    private const string UserSymbol = "<|user|>";
+    private const string AssistantSymbol = "<|assistant|>";
+    private const string EndSymbol = "<|end|>";
+    private const int SystemSymbolId = 32006;
+    private const int UserSymbolId = 32010;
+    private const int AssistantSymbolId = 32001;
+    private const int EndSymbolId = 32007;
+    private readonly Dictionary<string, int> _specialTokenMap = new Dictionary<string, int>
```
Comment (tarekgh) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs:22]: >_specialTokenMap [](http://example.com/codeflow?start=45&length=16)

Looks to me this not used.

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.ML.Tokenizers;
+
+/// <summary>
+/// The utility class to create tokenizer for phi-3 model.
+/// </summary>
+public class Phi3TokenizerHelper
```
Comment (tarekgh) [inline src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs:12]: >Phi3TokenizerHelper [](http://example.com/codeflow?start=13&length=19)

I suggest naming the cs file same as the class name.

**Comment by luisquintanilla** [comment]
@LittleLittleCloud samples look good. 

Something we might want to think about is the naming of the SK naming conventions for adding a model to make it more generic. 

Similar to:

https://github.com/feiyun0112/SemanticKernel.Connectors.OnnxRuntimeGenAI/blob/6efcbcfbb84c5ecdb5cc31215903b6df13f3f841/OnnxRuntimeGenAIServiceCollectionExtensions.cs#L26

File: src/Microsoft.ML.GenAI.Core/Pipeline/CausalLMPipeline.cs [hunk 6]
```diff
@@ -7,128 +7,266 @@
...
+        List<int[]> stopTokenIds = [[]];
+        if (stopSequences != null)
+        {
+            stopTokenIds.AddRange(stopSequences.Select(x =>
+            {
+                var tokens = this.Tokenizer.EncodeToTokens(x, out var _, false, false);
+
+                return tokens
+                .Where(t => t.Offset != (0, 0))
+                .Select(t => t.Id)
```
Comment (tarekgh) [inline src/Microsoft.ML.GenAI.Core/Pipeline/CausalLMPipeline.cs:253]: what this for?

Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Core/Pipeline/CausalLMPipeline.cs:253]: To detect the `_` token automatically added by tokenzier

Comment (tarekgh) [reply src/Microsoft.ML.GenAI.Core/Pipeline/CausalLMPipeline.cs:253]: ok, thanks. it is worth a comment. 

Also, EncodeToToken is very expensive comparing to EncodeToIds. If you are interested to optimize I can suggest some code to check for `_ ` token.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Core/Pipeline/CausalLMPipeline.cs:253]: Yes that would be very helpful

File: src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs (hunk 4 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.GenAI.Phi/Phi3/Phi3Tokenzier.cs:22]: Resolved

**Comment by LittleLittleCloud** [comment]
> @LittleLittleCloud samples look good.
> 
> Something we might want to think about is the naming of the SK naming conventions for adding a model to make it more generic.
> 
> Similar to:
> 
> https://github.com/feiyun0112/SemanticKernel.Connectors.OnnxRuntimeGenAI/blob/6efcbcfbb84c5ecdb5cc31215903b6df13f3f841/OnnxRuntimeGenAIServiceCollectionExtensions.cs#L26

Rename to `AddGenAIChatCompletion` and `AddGenAITextCompletion`

**Review by michaelgsharp — APPROVED** [review]

**Pull request merged**